server/integration/endpoint/src/main/java/org/infinispan/server/endpoint/subsystem/MultiTenantRouterService.java

```diff
@@ -0,0 +1,194 @@
+    private final java.util.Map<String, RestRouting> restRouting = new HashMap<>();
+
+    private final String name;
+    private final MultiTenantRouterConfigurationBuilder configurationBuilder;
+    private Optional<MultiTenantRouter> router = Optional.empty();
+
+    MultiTenantRouterService(MultiTenantRouterConfigurationBuilder configurationBuilder, Optional<String> serverName) {
+        this.name = constructServerName(serverName);
+        this.configurationBuilder = configurationBuilder;
+    }
+
+    private final String constructServerName(Optional<String> name) {
+        return name.orElse(DEFAULT_NAME);
+    }
+
+    @Override
+    public synchronized void start(final StartContext context) throws StartException {
+        ROOT_LOGGER.endpointStarting(name);
+        try {
+            SocketBinding hotrodSocketBinding = getHotrodSocketBinding().getOptionalValue();
+            InetSocketAddress hotrodAddress = hotrodSocketBinding != null ? hotrodSocketBinding.getSocketAddress() : null;
+
+            SocketBinding restSocketBinding = getRestSocketBinding().getOptionalValue();
+            InetSocketAddress restAddress = restSocketBinding != null ? restSocketBinding.getSocketAddress() : null;
+
+            if (hotrodAddress != null) {
+                HotRodRouterBuilder hotrodConfigurationBuilder = configurationBuilder.hotrod();
+                hotrodConfigurationBuilder.ip(hotrodAddress.getAddress());
+                hotrodConfigurationBuilder.port(hotrodAddress.getPort());
+            }
+            if (restAddress != null) {
+                configurationBuilder.rest().ip(restAddress.getAddress());
+                configurationBuilder.rest().port(restAddress.getPort());
+            }
+
+            hotrodRouting.forEach((sniHostName, routing) -> {
+                HotRodServer hotRod = routing.getHotRod().getValue();
+                SecurityRealm securityRealm = routing.getSecurityRealm().getValue();
+
+                SniNettyRouteSource source = new SniNettyRouteSource(sniHostName, securityRealm.getSSLContext());
+                NettyHandlerRouteDestination destination = new NettyHandlerRouteDestination(hotRod.getQualifiedName(), hotRod.getInitializer());
+
+                configurationBuilder.routing().add(new Route<>(source, destination));
+            });
+
+            restRouting.forEach((path, routing) -> {
+                Server restResource = routing.getRest().getValue().getServer().get();
+                String name = routing.getName();
+
+                RestRouteSource source = new RestRouteSource(path);
+                NettyRestServerRouteDestination destination = new NettyRestServerRouteDestination(name, restResource);
+
+                configurationBuilder.routing().add(new Route<>(source, destination));
+            });
+
+            this.router = Optional.of(new MultiTenantRouter(configurationBuilder.build()));
+            this.router.get().start();
```

Comment:
No no... it's correct here... 

Imagine a situation when an exception is thrown from `router#start` but for some reason Netty is running and it should be stopped (from `#stop` method). In order to do it we need `this.router` to have a reference to the router.